packages/json-api/src/-private/cache.ts

```diff
@@ -1661,14 +1676,18 @@ function getDefaultValue(
   }
 }
 
-function notifyAttributes(storeWrapper: CacheCapabilitiesManager, identifier: StableRecordIdentifier, keys?: string[]) {
+function notifyAttributes(
+  storeWrapper: CacheCapabilitiesManager,
+  identifier: StableRecordIdentifier,
+  keys?: Set<string>
+) {
   if (!keys) {
-    storeWrapper.notifyChange(identifier, 'attributes');
+    storeWrapper.notifyChange(identifier, 'attributes', null);
     return;
   }
 
-  for (let i = 0; i < keys.length; i++) {
-    storeWrapper.notifyChange(identifier, 'attributes', keys[i]);
+  for (const key of keys) {
+    storeWrapper.notifyChange(identifier, 'attributes', key);
```

Comment:
it may be worth adding a batch signature so we just forward these as a single batch. We already have the set and it would avoid N (record)*M (field) notifications in favor of just N.